fsspec/asyn.py

```diff
@@ -239,20 +239,38 @@ async def _run_coros_in_chunks(
         batch_size = len(coros)
 
     assert batch_size > 0
+
+    async def _run_coro(coro):
+        try:
+            return await asyncio.wait_for(coro, timeout=timeout)
+        except Exception as e:
+            if not return_exceptions:
+                raise
+            return e
+        finally:
+            callback.relative_update(1)
+
+    coros = iter(coros)
+    stop = False
+    pending = set()
     results = []
-    for start in range(0, len(coros), batch_size):
-        chunk = [
-            asyncio.Task(asyncio.wait_for(c, timeout=timeout))
-            for c in coros[start : start + batch_size]
-        ]
-        if callback is not DEFAULT_CALLBACK:
-            [
-                t.add_done_callback(lambda *_, **__: callback.relative_update(1))
-                for t in chunk
-            ]
-        results.extend(
-            await asyncio.gather(*chunk, return_exceptions=return_exceptions),
-        )
+
+    while pending or not stop:
+        while len(pending) < batch_size and not stop:
+            try:
+                coro = next(coros)
+            except StopIteration:
+                stop = True
+            else:
+                pending.add(asyncio.ensure_future(_run_coro(coro)))
+
+        if not pending:
+            break
+
+        done, pending = await asyncio.wait(pending, return_when=asyncio.FIRST_COMPLETED)
```

Comment:
It feels like this will alter the ordering of the results versus the inputs. Can you please try doing a `cat` with many files, to see if the order is preserved? I don't know that we have tests with genuinely significant numbers and latency.

Reply:
Ah, yes, I'm quite sure the order is not preserved. I think neither a queue-based approach nor this one would be able to preserve the order because coroutines are popped when done and appended to the results list. But since the coroutines input is a list and not a generator, we know its length. So, we can allocate a same-length `results` list and insert the results at the same index of the corresponding coroutine.

Let me add this quickly for you to review.

Reply:
Mh, but list insertions at some index have complexity O(n). :confused:

Reply:
Done.

Reply:
So you would need a dict, assuming the inputs or coroutines are hashable

Reply:
Sorry, I've updated by comment in the meantime. Setting items in lists is O(1), so all is good.

Reply:
A bit concerned this may still be inefficient in scheduling the workers. IE: it waits until the batch is completed before the next batch is scheduled: https://github.com/fsspec/filesystem_spec/pull/1544#discussion_r1526491315

Reply:
I don't think this happens here becausw of `return_when=asyncio.FIRST_CIMPLETED`.

Reply:
When I print `len(pending)` like this

```diff
     pending = set()
 
     while pending or i < n:
+        print("DEBUG: len(pending) =", len(pending))
         while len(pending) < batch_size and i < n:
             pending.add(asyncio.ensure_future(_run_coro(coros[i], i)))
             i += 1
```

and modify the `test_run_coros_in_chunks` test in `tests/test_async.py` like this

```diff
     async def runner():
         nonlocal total_running
+        import random
 
         total_running += 1
-        await asyncio.sleep(0)
+        await asyncio.sleep(random.uniform(0.0, 1.0))
         if total_running > 4:
             raise ValueError("More than 4 coroutines are running together")
         total_running -= 1
```

then the printed info is this for me:

<details>

<summary>Log</summary>

```
DEBUG: len(pending) = 0
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 2
DEBUG: len(pending) = 1
DEBUG: len(pending) = 0
DEBUG: len(pending) = 0
DEBUG: len(pending) = 0
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 2
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 2
DEBUG: len(pending) = 1
DEBUG: len(pending) = 0
DEBUG: len(pending) = 0
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 2
DEBUG: len(pending) = 1
DEBUG: len(pending) = 0
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 3
DEBUG: len(pending) = 2
DEBUG: len(pending) = 1
```

</details>

I believe it works as we expect. WDYT, @martindurant?